{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "WBT-MBT2-25E <i>Programming Python for Bioinformatics</i> &copy; 2020-2025 Michal Bukowski (m.bukowski@uj.edu.pl) Department of Analytical Biochemistry, Faculty of Biochemistry, Biophysics and Biotechnology, Jagiellonian University\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddeeff;border-radius:15px\">DNASeq, a new data type for processing DNA sequences</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=3>\n",
    "    \n",
    "Below you will find an incomplete implementation of the class `DNASeq` that is supposed to define a new type of a data structure facilitating different operations performed on DNA sequences. These include operations that might be used to process multiple sequence alignments in FASTA format, which apart from letters belonging to the [IUPAC ambiguous DNA alphabet](https://www.bioinformatics.org/sms/iupac.html), may contain gaps (the minus sign `-`).<br>\n",
    "\n",
    "Complete gradually the implementation of the class DNASeq (<u>it must be contained in one cell</u>). Follow the comments in the code. Upon completion of each numbered part, run code in the last section `Test the code!`. It will help you to build the class step by step.<br>\n",
    "\n",
    "<u>Remove the comment sign</u> `#` placed before subsequent code lines (manually or using `Ctrl + /` for multiple lines at once) and fill in the gaps marked with `???`. Sometimes it will require only one line of code or a simple expression, sometimes a whole block of code.<br>\n",
    "\n",
    "Once again, remember to fill the code step by step, part by part.<br>\n",
    "    \n",
    "<b>REMEMBER</b> to <u>rerun</u> the cell with DNASeq class definition after making any changes. Otherwise, those will not take any effect.<br>\n",
    "    \n",
    "<b>TEST</b> your code upon finishing each section. It <u>must work</u> before you move forward. In order to make the code work, you <u>must implement sections in the given order</u>.\n",
    "\n",
    "</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#eeddff;border-radius:15px\">Submission of the task solution</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=3>\n",
    "    \n",
    "<b>Bioinformatics Students</b>. Having the `DNASeq` class implemented, prepare it for submission to [TestPyPI](https://test.pypi.org/). Follow the procedure presented during the lecture and described in the corresponding presentation. Apart from the code file with the `DNASeq` class implementation, you will need to prepare a basic `README.md` file (in the Markdown format). Shortly describe in it main features of the class and give a few working examples of its use. A demo `README.md` file is published alongside materials form the lecture. Name your package <b>seqlib<span style='color:#cc0000'>FiLa</span></b>, where <b><span style='color:#cc0000'>FiLa</span></b> are your initials, eg. <b>seqlib<span style='color:#cc0000'>MaSk</span></b> if you are Maria Sk≈Çodowska-Curie. Once you have the package ready, publish it on [GitHub](https://github.com/). Remember to provide proper links to the GitHub repository in the `pyproject.toml` file and submit the whole package to [TestPyPI](https://test.pypi.org/). Once you have everything completed, <u><span style='color:#229900'>submit the link to your GitHub repo in PP4B Team Assignments section</span></u>.\n",
    "    \n",
    "<b>Molecular Biotechnology and Erasmus Students</b>. Complete and <u><span style='color:#229900'>submit this Jupyter notebook in PP4B Team Assignments section</span></u> (in its native `*.ipynb` format). Nevertheless, feel encouraged to follow the requirements for Bioinformatics Students, i.e. package your code and publish on [GitHub](https://github.com/) and [TestPyPI](https://test.pypi.org/) as described in the previous paragraph.\n",
    "\n",
    "</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### On the GenBank notation\n",
    "#### <u>An example for positions/coordinates/indices 1 and 4</u>:\n",
    "(<i>start</i> and <i>stop</i> in Python, <i>beginning/start</i> and <i>end</i> in GenBank)\n",
    "<pre style=\"font-family:Monospace\">\n",
    "<span style=\"color:#00aaaa\">          |----&gt;|</span>\n",
    "<span style=\"color:#aaaaaa\"> Python: 0 1 2 3 4 5 6 7 8</span>\n",
    "         A T C G T A C G A\n",
    "<span style=\"color:#aaaaaa\">GenBank: 1 2 3 4 5 6 7 8 9</span>\n",
    "<span style=\"color:#00aaaa\">        |------&gt;|</span>\n",
    "         T A G C A T G C T   (complementary/minus strand)\n",
    "<span style=\"color:#aa0000\">        |&lt;------|</span>\n",
    "\n",
    "<span style=\"color:#aaaaaa\"> Python:</span> seq[5]   <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'A'</span>\n",
    "<span style=\"color:#aaaaaa\">GenBank:</span> seq[5]   <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'T'</span>\n",
    "\n",
    "<span style=\"color:#aaaaaa\"> Python:</span> seq[1:4] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'TCG'</span>\n",
    "<span style=\"color:#aaaaaa\">GenBank:</span> seq[1:4] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#00aaaa\">'ATCG'</span>\n",
    "\n",
    "<span style=\"color:#aaaaaa\"> Python:</span> seq[4:1] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#aa0000\">''</span>\n",
    "<span style=\"color:#aaaaaa\">GenBank:</span> seq[4:1] <span style=\"color:#aaaaaa\">-&gt;</span> <span style=\"color:#aa0000\">'CGAT'</span>   (5' to 3' direction, reverse-complement)\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You will use numpy module,\n",
    "# TODO: import numpy under the name np.\n",
    "\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The overview of the implemented class\n",
    "```Python\n",
    "class DNASeq:\n",
    "\n",
    "    ALPH\n",
    "    \n",
    "    __init__()\n",
    "    __repr__()\n",
    "    __str__()\n",
    "    __len__()\n",
    "    __add__()\n",
    "    __getitem__()\n",
    "    \n",
    "    from_file()\n",
    "    revcmpl()\n",
    "    copy()\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# And there it is, our DNASeq class!\n",
    "class DNASeq:\n",
    " \n",
    "    #===SECTION=1========================================================\n",
    "    \n",
    "    # A dictionary with the IUPAC ambiguous DNA alphabet\n",
    "    # (https://www.bioinformatics.org/sms/iupac.html)\n",
    "    # and a gap sign. Each key is paired with a value\n",
    "    # that is the key's complement. This dictionary\n",
    "    # is defined as a class variable. Revise\n",
    "    # how to access values of such variables,\n",
    "    # you will need it in the coming sections.\n",
    "    # TODO: Fill up the missing letters,\n",
    "    #       make them capitals.\n",
    "    \n",
    "    ALPH = {\n",
    "        'A' : 'T',   'T' : 'A',   'G' : 'C',   'C' : 'G',\n",
    "        'K' : 'M',   'M' : 'K',   'B' : 'V',   'D' : 'H',\n",
    "        'H' : 'D',   'V' : 'B',   'N' : 'N',   '-' : '-'\n",
    "    }\n",
    "    \n",
    "    # TEST the code in the Section 1.\n",
    "    \n",
    "    \n",
    "    #===SECTION=2========================================================\n",
    "    \n",
    "    # The special method __init__() initialising the\n",
    "    # initial state of a newly created object (a class instance).\n",
    "    # TODO: Define the special method __init__() proper for\n",
    "    #       initialisation of a new object of DNASeq type,\n",
    "    #       which is to be created as following:\n",
    "    #       seq = DNASeq(seqid, title, seq)\n",
    "    \n",
    "    def __init__(self, seqid, title, seq):\n",
    "        # TODO: Assign values of the arguments passed\n",
    "        #       while creating a new object to the\n",
    "        #       object attributes that will have\n",
    "        #       the same names.\n",
    "        \n",
    "        self.seqid = seqid\n",
    "        self.title = title\n",
    "        self.seq = seq\n",
    "        \n",
    "    # TEST the code in the Section 2.\n",
    "        \n",
    "        \n",
    "    #===SECTION=3========================================================\n",
    "        \n",
    "    # Now we need a class method that will help us to\n",
    "    # deserialise sequences from a FASTA format file to\n",
    "    # a collection of DNASeq objects.\n",
    "    # TODO: Define a class method from_file() that next to\n",
    "    #       the automatically passed reference to a class\n",
    "    #       will accept an argument with a file path, and\n",
    "    #       will work as following:\n",
    "    #       seqs = DNASeq.from_file('some/path/some_file.fasta')\n",
    "    \n",
    "    @classmethod\n",
    "    def from_file(cls, filepath):\n",
    "        # TODO: Create an empty seqs dictionary and a variable for\n",
    "        #       a current seqid.\n",
    "        seqs, seqid = {}, None\n",
    "        \n",
    "        # TODO: Open the input file for reading\n",
    "        with open(filepath, 'r') as f:\n",
    "            \n",
    "        # TODO: Start iteratively parsing the opened file\n",
    "        #       line by line. Refer to the lecture presentationchay\n",
    "        #       for details.\n",
    "            for line in f:\n",
    "            # TODO: Parse sequences form the input file.\n",
    "                if line[0] == '>':\n",
    "                    header = line[1:-1].split(' ',1)\n",
    "                    seqid = header[0]\n",
    "                    title = header[1] if len(header) > 1 else ''\n",
    "                    if seqid in seqs:\n",
    "                        raise Exception(f'Non-unige: {seqid}')\n",
    "                    seqs[seqid] = cls(seqid, title, [])\n",
    "                else:\n",
    "                    seqs[seqid].seq.append(line[:-1])\n",
    "            for seq in seqs.values():\n",
    "                seq.seq = ''.join(seq.seq)\n",
    "         \n",
    "            # TODO: An extra step: convert every sequence line\n",
    "            #       to a set and use a proper method for Python sets,\n",
    "            #       which will allow you to see if the\n",
    "            #       alphabet of seq fragment is a subset of ALPHABET,\n",
    "            #       which is one of previously defined\n",
    "            #       class variables. If not raise an exception.\n",
    "                if not set(seq.seq).issubset(cls.ALPH.keys()):\n",
    "                    raise ValueError(f\"Invalid characters found in sequence: {seq.seq}\")\n",
    "            \n",
    "        # TODO: Return the reference to the collection of new objects\n",
    "        #       of the cls class.\n",
    "        f.close() \n",
    "        return seqs \n",
    "    \n",
    "    # TEST the code in the Section 3.\n",
    "    \n",
    "        \n",
    "    #===SECTION=4========================================================\n",
    "    \n",
    "    # The special method __repr__() returns a string representation of\n",
    "    # an object, we will define this representation as 10 first letters\n",
    "    # of the sequence stored within an object and three dots '...'\n",
    "    # UNLESS the sequence length is shorter or equal to 10 letters.\n",
    "    \n",
    "    def __repr__(self):\n",
    "        if len(self.seq) <= 10:\n",
    "            return self.seq\n",
    "        else:\n",
    "            return self.seq[:10] + '...'\n",
    "    \n",
    "    # TEST the code in the Section 4.\n",
    "    \n",
    "    \n",
    "    #===SECTION=5========================================================\n",
    "        \n",
    "    # TODO: Define the proper special method that will return the length of\n",
    "    #       the sequence contained within a given DNASeq object when the reference\n",
    "    #       to that object is passed to the built-in len() function.\n",
    "    \n",
    "    def __len__(self):\n",
    "        return len(self.seq)\n",
    "    \n",
    "    # TEST the code in the Section 5.\n",
    "    \n",
    "    \n",
    "    #===SECTION=6========================================================\n",
    "    \n",
    "    # TODO: Implement the special method __str__(), which returns\n",
    "    #       a string whenever a DNASeq object is being converted to one,\n",
    "    #       eg. str(obj) or print(obj).\n",
    "    #       Make it return the sequence contained within an object \n",
    "    #       as FASTA formatted sequence string.\n",
    "    #       Let the sequence be divided into 60-character lines.\n",
    "    \n",
    "    def __str__(self):\n",
    "        lines = '\\n'.join([self.seq[i:i+60] for i in range(0, len(self.seq), 60)])\n",
    "\n",
    "        title = f' {self.title}' if self.title != '' else ''\n",
    "\n",
    "        fasta = f'>{self.seqid}{title}\\n{lines}'\n",
    "\n",
    "        return fasta\n",
    "    \n",
    "    # TEST the code in the Section 6.\n",
    "    \n",
    "    \n",
    "    #===SECTION=7=======================================================\n",
    "    \n",
    "    # Let's create a custom method and call it revcmpl().\n",
    "    # It will simply return a new object of DNASeq type, which\n",
    "    # will contain a reverse-complement sequence to the\n",
    "    # one contained within the original object the method is called from.\n",
    "    # To make both object distinguishable, create a new\n",
    "    # seqid for the sequence by adding '_revcmpl' suffix\n",
    "    # to the seqid of the original one.\n",
    "    \n",
    "    def revcmpl(self):\n",
    "        # TODO: Using string slices and step equal to -1,\n",
    "        # reverse the sequence.\n",
    "        \n",
    "        seq = self.seq[::-1]\n",
    "        \n",
    "        # TODO: Using string method join(), the class dictionary ALPH and a\n",
    "        #       list comprehension expression, translate the reversed sequence and\n",
    "        #       convert into a string\n",
    "        \n",
    "        seq_revcmpl = ''.join([self.ALPH[base] for base in seq])\n",
    "        \n",
    "        # TODO: Create seqid variable and assign to it the object's seqid\n",
    "        #       together with the suffix '_revcmpl'.\n",
    "        \n",
    "        seqid = self.seqid + '_revcmpl'\n",
    "        \n",
    "        # TODO: Create a new object of the DNASeq type using the new seqid,\n",
    "        #       title contained in the object as well as\n",
    "        #       reversed and translated sequence, return the reference\n",
    "        #       to new object.\n",
    "        \n",
    "        revcmpl_obj = DNASeq(seqid, self.title, seq_revcmpl)\n",
    "        return revcmpl_obj\n",
    "    \n",
    "    # TEST the code in the Section 7.\n",
    "    \n",
    "    \n",
    "    #===SECTION=8========================================================\n",
    "    \n",
    "    # Beware! This is the hardest part to go through.\n",
    "    # We will implement the special method __getitem__()\n",
    "    # that allows to program what happens when an object\n",
    "    # is indexed or sliced (like a list or a string),\n",
    "    # eg. obj[3] or obj[4:5].\n",
    "    #\n",
    "    # We know that indexing and slicing in Python works as follows:\n",
    "    # - indexing starts at 0\n",
    "    # - when a slice is taken [3:4] the first index is included\n",
    "    #   the last not: eg. s = 'abcde', s[1:3] -> 'bc'\n",
    "    #                          01234\n",
    "    # - slice [3:1] will be an empty sequence\n",
    "    #\n",
    "    # When it comes to biological sequences, those are indexed\n",
    "    # according to GenBank notation, which is a bit different:\n",
    "    # - indexing starts at 1\n",
    "    # - when a fragment is requested, both indices are inclusive:\n",
    "    #   seq = 'ATGCTACG', seq[1:3] -> 'ATG'\n",
    "    #          12345678\n",
    "    # - the start index greater than stop index indicates\n",
    "    #   a reverse complement (the complementary strand):\n",
    "    #   seq[3:1] -> 'CAT'\n",
    "    #\n",
    "    # Now we will try to implement this behaviour in case of\n",
    "    # our objects of DNASeq type.\n",
    "    \n",
    "    def __getitem__(self, key):\n",
    "        \n",
    "        if isinstance(key, slice):\n",
    "            # If the key is a slice object, it has three properties:\n",
    "            # start, stop and step, eg. list[start:stop:step].\n",
    "            # If any is missing its value is None, eg. list[start:end].\n",
    "            # First two will be equivalent of beginning and end\n",
    "            # in GenBank notation.\n",
    "            \n",
    "            # The whole point here is to translate GenBank indices\n",
    "            # into Python ones and index or slice the sequence\n",
    "            # contained in the DNASeq object, which is surely of\n",
    "            # Python string format, so it cannot be directly indexed\n",
    "            # or sliced with GenBank indices without translating them.\n",
    "            \n",
    "            # To make the code nicer, let's assign values of slice\n",
    "            # properties to single variables.\n",
    "            \n",
    "            start, end, step = key.start, key.stop, key.step\n",
    "            \n",
    "            # Let's test a few possibilities and exclude those\n",
    "            # that cannot be translated into GenBank notation.\n",
    "            \n",
    "            if step is not None:\n",
    "                # There is not an equivalent of step in GenBank notation.\n",
    "                # If step is defined anyway, raise a KeyError.\n",
    "                \n",
    "                raise KeyError('Step is not allowed in GenBank notation')\n",
    "            \n",
    "            if start is None:\n",
    "                # Start must be provided.\n",
    "                \n",
    "                raise KeyError('start index is required')\n",
    "            \n",
    "            if end is None:\n",
    "                # As well as the end.\n",
    "                \n",
    "                raise KeyError('end index is required')\n",
    "                \n",
    "            if not np.issubdtype(type(start), np.integer) or \\\n",
    "               not np.issubdtype(type(end), np.integer):\n",
    "                # Start and end must be defined as integer values,\n",
    "                # otherwise raise a KeyError.\n",
    "                \n",
    "                raise TypeError('Start and end must be integers')\n",
    "            \n",
    "            if start <= 0 or end <= 0:\n",
    "                # Both must be greater than 0 as in GenBank notation\n",
    "                # indexing starts at 1.\n",
    "                \n",
    "                raise KeyError('Minimal value for start and end is 1')\n",
    "                \n",
    "            # Now we are sure there is only start and end in the slice,\n",
    "            # and that both are integer type values and equal or greater to 1.\n",
    "            # Let's move on then.\n",
    "            \n",
    "            # TODO: Create a variable strand and set it to 1 if start is\n",
    "            #       less or equal to end, otherwise to -1.\n",
    "                \n",
    "            strand = 1 if start <= end else -1\n",
    "            \n",
    "            # TODO: If strand is equal to -1, swap the values of start and end,\n",
    "            #       by using tuple unpacking, so start is less than end (they are ordered).\n",
    "            #       You may also do it without actually checking whether one is grater\n",
    "            #       than the other, using build-in sorted() of numpy np.sort() function and\n",
    "            #       tuple unpacking expression.\n",
    "\n",
    "            if strand == -1:\n",
    "                start, end = np.sort([start, end])\n",
    "            \n",
    "            # TODO: Create a new seqid by adding to the existing one\n",
    "            #       the suffix '_loc(start_end)', where start and end are\n",
    "            #       values of our variables.\n",
    "            \n",
    "            seqid = f\"{self.seqid}_loc({start}_{end})\"\n",
    "            \n",
    "            # TODO: Decrease start by 1. If start in GenBank notation is 1,\n",
    "            #       it is and equivalent of 0 in Python indexing (etc.),\n",
    "            #       so it needs to be decreased by 1 to be translated\n",
    "            #       from GenBank to Python.\n",
    "            #\n",
    "            #       You must NOT decrease end, as in GenBank it is inclusive,\n",
    "            #       but in Python exclusive, so it is decreased\n",
    "            #       somewhat automatically.\n",
    "            \n",
    "            start -= 1\n",
    "            \n",
    "            # TODO: Create a new object of DNASeq type by using new seqid\n",
    "            #       you created a few lines before, title of the existing object,\n",
    "            #       a slice of the sequence contained in that object by using\n",
    "            #       the adjusted start index and the end index (string slicing).\n",
    "            \n",
    "            sub_seq = DNASeq(seqid, self.title, self.seq[start:end])\n",
    "            \n",
    "            # TODO: If strand is -1, assign to the same reference (\"variable\")\n",
    "            #       a reverse complement of the new sequence object by using\n",
    "            #       the method revcmpl() you implemented before.\n",
    "\n",
    "            if strand == -1:\n",
    "                sub_seq = sub_seq.revcmpl()\n",
    "            \n",
    "                \n",
    "            # TODO: return the reference to the new sequence object.\n",
    "                \n",
    "            return sub_seq\n",
    "                \n",
    "        else:\n",
    "            \n",
    "            # If we are here, it means that key is not a slice object.\n",
    "            # Then we allow it be an integer greater than 0,\n",
    "            # compliant with the GenBank notation.\n",
    "            \n",
    "            if not np.issubdtype(type(key), np.integer):\n",
    "                raise TypeError('Index must be an integer')\n",
    "                \n",
    "            if key <= 0:\n",
    "                raise KeyError('Minimal value of index is 1')\n",
    "                \n",
    "            # In case it is just one letter not a slice, we\n",
    "            # will return simply a letter from the sequence\n",
    "            # (a string value), NOT a new DNASeq object.\n",
    "            # We need to remember to decrease the key value\n",
    "            # by one to translate it from GenBank to Python.\n",
    "                \n",
    "            return self.seq[key - 1]\n",
    "    \n",
    "    # TEST the code in the Section 8.\n",
    "        \n",
    "        \n",
    "    #===SECTION=9========================================================\n",
    "    \n",
    "    # Implement the special method __add__(), which will be\n",
    "    # invoked when two objects of the DNASeq type are being added,\n",
    "    # eg. seq_3 = seq_1 + seq_2\n",
    "    #\n",
    "    # Make the result of addition a new object of type(self) type\n",
    "    # with its properties values as follows:\n",
    "    # - seqid will be seqids of added objects,\n",
    "    #   separated by an underscore\n",
    "    # - title will be titles of added objects,\n",
    "    #   separated by an underscore\n",
    "    # - seq will simply be a concatenation of both sequences\n",
    "    #   stored within the added objects\n",
    "    #\n",
    "    # return the reference to the new object.\n",
    "\n",
    "    def __add__(self, other):\n",
    "        new_seqid = f\"{self.seqid}_{other.seqid}\"\n",
    "        new_title = f\"{self.title}_{other.title}\"\n",
    "        new_seq = self.seq + other.seq\n",
    "        \n",
    "        return DNASeq(new_seqid, new_title, new_seq)\n",
    "    \n",
    "    \n",
    "    # TEST the code in the Section 9.\n",
    "    \n",
    "    \n",
    "    #===SECTION=10========================================================\n",
    "        \n",
    "    # Define a custom method copy() that will return an exact copy\n",
    "    # of the existing object (as a type(self) type object).\n",
    "        \n",
    "    def copy(self):\n",
    "        return DNASeq(self.seqid, self.title, self.seq)\n",
    "    \n",
    "    # TEST the code in the Section 10.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddeeff;border-radius:15px\">Test the code!</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 1<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "The whole dictionary: {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'K': 'M', 'M': 'K', 'B': 'V', 'D': 'H', 'H': 'D', 'V': 'B', 'N': 'N', '-': '-'}<br><br>\n",
    "A letter complementary to \"C\": G\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The whole dictionary: {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'K': 'M', 'M': 'K', 'B': 'V', 'D': 'H', 'H': 'D', 'V': 'B', 'N': 'N', '-': '-'} \n",
      "\n",
      "A letter complementary to \"C\": G \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Let's see our class variables:\n",
    "print('The whole dictionary:', DNASeq.ALPH, '\\n')\n",
    "\n",
    "print('A letter complementary to \"C\":', DNASeq.ALPH['C'], '\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 2<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "new_seq (an exemplary sequence): ATCGT...\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "new_seq (an exemplary sequence): ATCGT...\n"
     ]
    }
   ],
   "source": [
    "seq = DNASeq('new_seq', 'an exemplary sequence',\n",
    "             'ATCGTAGGATCGGATTAGAGCGATTAGCTAG')\n",
    "\n",
    "print(f'{seq.seqid} ({seq.title}): {seq.seq[:5]}...')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 3<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "arcC, aroE, glpF, gmk, pta, tpi, yqiL<br><br>\n",
    "yqiL (Acetyle coenzyme A acetyltransferase): GCGTT...\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arcC, aroE, glpF, gmk, pta, tpi, yqiL\n",
      "\n",
      "yqiL (Acetyle coenzyme A acetyltransferase): GCGTT...\n"
     ]
    }
   ],
   "source": [
    "# Deserialise sequences from a FASTA formated file\n",
    "# into a collection of objects of DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "seqids = ', '.join( seqs.keys() )\n",
    "\n",
    "print(seqids)\n",
    "\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "print(f'\\n{seq.seqid} ({seq.title}): {seq.seq[:5]}...')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 4<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "{'arcC': TTATTAATCC...,<br>\n",
    " 'aroE': AATTTTAATT...,<br>\n",
    " 'glpF': GGTGCTGATT...,<br>\n",
    " 'gmk': CGAATATTTG...,<br>\n",
    " 'pta': GCAACACAAT...,<br>\n",
    " 'tpi': CACGAAACAG...,<br>\n",
    " 'yqiL': GCGTTTAAAG...}\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'arcC': TTATTAATCC...,\n",
       " 'aroE': AATTTTAATT...,\n",
       " 'glpF': GGTGCTGATT...,\n",
       " 'gmk': CGAATATTTG...,\n",
       " 'pta': GCAACACAAT...,\n",
       " 'tpi': CACGAAACAG...,\n",
       " 'yqiL': GCGTTTAAAG...}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class.\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "seqs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 5<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "516\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "516"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# Select one of the sequences by its sequence id (seqid):\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "# Look up the length of the contained sequence:\n",
    "len(seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 6<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;yqiL Acetyle coenzyme A acetyltransferase<br>\n",
    "GCGTTTAAAGACGTGCCAGCCTATGATTTAGGTGCGACTTTAATAGAACATATTATTAAA<br>\n",
    "GAGACGGGTTTGAATCCAAGTGAGATTGATGAAGTTATCATCGGTAACGTACTACAAGCA<br>\n",
    "GGACAAGGACAAAATCCAGCACGAATTGCTGCTATGAAAGGTGGCTTGCCAGAAACAGTA<br>\n",
    "CCTGCATTTACAGTGAATAAAGTATGTGGTTCTGGGTTAAAGTCGATTCAATTAGCATAT<br>\n",
    "CAATCTATTGTGACTGGTGAAAATGACATCGTGCTAGCTGGCGGTATGGAGAATATGTCT<br>\n",
    "CAGTCACCAATGCTTGTCAACAACAGTCGCTTCGGTTTTAAAATGGGACATCAATCAATG<br>\n",
    "GTTGATAGCATGGTATATGATGGTTTAACAGATGTATTTAATCAATATCATATGGGTATT<br>\n",
    "ACTGCTGAAAATTTAGTGGAGCAATATGGTATTTCAAGAGAAGAACAAGATACATTTGCT<br>\n",
    "GTAAACTCACAACAAAAAGCAGTACGTGCACAGCAA\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">yqiL Acetyle coenzyme A acetyltransferase\n",
      "GCGTTTAAAGACGTGCCAGCCTATGATTTAGGTGCGACTTTAATAGAACATATTATTAAA\n",
      "GAGACGGGTTTGAATCCAAGTGAGATTGATGAAGTTATCATCGGTAACGTACTACAAGCA\n",
      "GGACAAGGACAAAATCCAGCACGAATTGCTGCTATGAAAGGTGGCTTGCCAGAAACAGTA\n",
      "CCTGCATTTACAGTGAATAAAGTATGTGGTTCTGGGTTAAAGTCGATTCAATTAGCATAT\n",
      "CAATCTATTGTGACTGGTGAAAATGACATCGTGCTAGCTGGCGGTATGGAGAATATGTCT\n",
      "CAGTCACCAATGCTTGTCAACAACAGTCGCTTCGGTTTTAAAATGGGACATCAATCAATG\n",
      "GTTGATAGCATGGTATATGATGGTTTAACAGATGTATTTAATCAATATCATATGGGTATT\n",
      "ACTGCTGAAAATTTAGTGGAGCAATATGGTATTTCAAGAGAAGAACAAGATACATTTGCT\n",
      "GTAAACTCACAACAAAAAGCAGTACGTGCACAGCAA\n"
     ]
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# Select one of the sequences by its sequence id (seqid):\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "# Print it as if it was a string value:\n",
    "print(seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 7<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;yqiL_revcmpl Acetyle coenzyme A acetyltransferase<br>\n",
    "TTGCTGTGCACGTACTGCTTTTTGTTGTGAGTTTACAGCAAATGTATCTTGTTCTTCTCT<br>\n",
    "TGAAATACCATATTGCTCCACTAAATTTTCAGCAGTAATACCCATATGATATTGATTAAA<br>\n",
    "TACATCTGTTAAACCATCATATACCATGCTATCAACCATTGATTGATGTCCCATTTTAAA<br>\n",
    "ACCGAAGCGACTGTTGTTGACAAGCATTGGTGACTGAGACATATTCTCCATACCGCCAGC<br>\n",
    "TAGCACGATGTCATTTTCACCAGTCACAATAGATTGATATGCTAATTGAATCGACTTTAA<br>\n",
    "CCCAGAACCACATACTTTATTCACTGTAAATGCAGGTACTGTTTCTGGCAAGCCACCTTT<br>\n",
    "CATAGCAGCAATTCGTGCTGGATTTTGTCCTTGTCCTGCTTGTAGTACGTTACCGATGAT<br>\n",
    "AACTTCATCAATCTCACTTGGATTCAAACCCGTCTCTTTAATAATATGTTCTATTAAAGT<br>\n",
    "CGCACCTAAATCATAGGCTGGCACGTCTTTAAACGC\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">yqiL_revcmpl Acetyle coenzyme A acetyltransferase\n",
      "TTGCTGTGCACGTACTGCTTTTTGTTGTGAGTTTACAGCAAATGTATCTTGTTCTTCTCT\n",
      "TGAAATACCATATTGCTCCACTAAATTTTCAGCAGTAATACCCATATGATATTGATTAAA\n",
      "TACATCTGTTAAACCATCATATACCATGCTATCAACCATTGATTGATGTCCCATTTTAAA\n",
      "ACCGAAGCGACTGTTGTTGACAAGCATTGGTGACTGAGACATATTCTCCATACCGCCAGC\n",
      "TAGCACGATGTCATTTTCACCAGTCACAATAGATTGATATGCTAATTGAATCGACTTTAA\n",
      "CCCAGAACCACATACTTTATTCACTGTAAATGCAGGTACTGTTTCTGGCAAGCCACCTTT\n",
      "CATAGCAGCAATTCGTGCTGGATTTTGTCCTTGTCCTGCTTGTAGTACGTTACCGATGAT\n",
      "AACTTCATCAATCTCACTTGGATTCAAACCCGTCTCTTTAATAATATGTTCTATTAAAGT\n",
      "CGCACCTAAATCATAGGCTGGCACGTCTTTAAACGC\n"
     ]
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# Select one of the sequences by its sequence id (seqid):\n",
    "seq = seqs['yqiL']\n",
    "\n",
    "# Get its reverse complement:\n",
    "new_seq = seq.revcmpl()\n",
    "\n",
    "# Print it as if it was a string value:\n",
    "print( new_seq )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 8<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;gmk Guanylate kinase<br>\n",
    "CGAATATTTGAAGATCCAAGTACATCATATAAGTATTCTATTTCAATGACAACACGTCAA<br>\n",
    "ATGCGTGAAGGTGAAGTTGATGGCGTAGATTACTTTTTTAAAACTAGGGATGCGTTTGAA<br>\n",
    "GCTTTAATCAAAGATGACCAATTTATAGAATATGCTGAATATGTAGGCAACTATTATGGT<br>\n",
    "ACACCAGTTCAATATGTTAAAGATACAATGGACGAAGGTCATGATGTATTTTTAGAAATT<br>\n",
    "GAAGTAGAAGGTGCAAAGCAAGTTAGAAAGAAATTTCCAGATGCGCTATTTATTTTCTTA<br>\n",
    "GCACCTCCAAGTTTAGAACACTTGAGAGAGCGATTAGTAGGTAGAGGAACAGAATCTGAT<br>\n",
    "GAGAAAATACAAAGTCGTATTAACGAAGCGCGTAAAGAAGTTGAAATGATGAATTTA\n",
    "</p>\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:10pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "and:\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&lt;class 'str'&gt;<br>\n",
    "C<br><br>\n",
    "&lt;class 'str'&gt;<br>\n",
    "ATATTT<br><br>\n",
    "&lt;class 'str'&gt;<br>\n",
    "<br>\n",
    "</p>\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:10pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "and:\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&lt;class 'str'&gt;<br>\n",
    "C<br><br>\n",
    "&lt;class '__main__.DNASeq'&gt;<br>\n",
    "&gt;gmk_loc(4_9) Guanylate kinase<br>\n",
    "ATATTT<br><br>\n",
    "&lt;class '__main__.DNASeq'&gt;<br>\n",
    "&gt;gmk_loc(1_1) Guanylate kinase<br>\n",
    "C\n",
    "</p>\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:10pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "and:\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;gmk_loc(4_9) Guanylate kinase<br>\n",
    "ATATTT<br><br>\n",
    "&gt;gmk_loc(4_9)_revcmpl Guanylate kinase<br>\n",
    "AAATAT<br><br>\n",
    "&gt;gmk_loc(4_9)_revcmpl Guanylate kinase<br>\n",
    "AAATAT\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gmk Guanylate kinase\n",
      "CGAATATTTGAAGATCCAAGTACATCATATAAGTATTCTATTTCAATGACAACACGTCAA\n",
      "ATGCGTGAAGGTGAAGTTGATGGCGTAGATTACTTTTTTAAAACTAGGGATGCGTTTGAA\n",
      "GCTTTAATCAAAGATGACCAATTTATAGAATATGCTGAATATGTAGGCAACTATTATGGT\n",
      "ACACCAGTTCAATATGTTAAAGATACAATGGACGAAGGTCATGATGTATTTTTAGAAATT\n",
      "GAAGTAGAAGGTGCAAAGCAAGTTAGAAAGAAATTTCCAGATGCGCTATTTATTTTCTTA\n",
      "GCACCTCCAAGTTTAGAACACTTGAGAGAGCGATTAGTAGGTAGAGGAACAGAATCTGAT\n",
      "GAGAAAATACAAAGTCGTATTAACGAAGCGCGTAAAGAAGTTGAAATGATGAATTTA\n"
     ]
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# Select one of the sequences by its sequence id (seqid):\n",
    "seq = seqs['gmk']\n",
    "\n",
    "# Print it as if it was a string value:\n",
    "print(seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "C\n",
      "\n",
      "<class 'str'>\n",
      "ATATTT\n",
      "<class 'str'>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# In Python string indexing starts at 0.\n",
    "# In the case of slices, the other index is exclusive.\n",
    "# You can see it by indexing a sequence stored in\n",
    "# a DNASeq object directly.\n",
    "\n",
    "# Get a letter from the sequence (string variable) at index 0.\n",
    "letter = seq.seq[0]\n",
    "\n",
    "# Print the type of the returned object and the letter itself:\n",
    "print(type(letter), letter, sep='\\n', end='\\n\\n')\n",
    "\n",
    "# Get a slice of the sequence (string variable):\n",
    "s = seq.seq[3:9]\n",
    "\n",
    "# Print the type of the returned object and its content:\n",
    "print(type(s), s, sep='\\n')\n",
    "\n",
    "# Get a reversed slice of the sequence (string variable):\n",
    "s = seq.seq[9:3]\n",
    "\n",
    "# Print the type of the returned object and its content:\n",
    "print(type(s), s, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "C\n",
      "\n",
      "<class '__main__.DNASeq'>\n",
      ">gmk_loc(4_9) Guanylate kinase\n",
      "ATATTT\n",
      "\n",
      "<class '__main__.DNASeq'>\n",
      ">gmk_loc(1_1) Guanylate kinase\n",
      "C\n"
     ]
    }
   ],
   "source": [
    "# Thanks to implementing the special method __getitem__,\n",
    "# you can index a DNASeq object directly, relying on the\n",
    "# GenBank notation. It means that indexing starts at 1\n",
    "# and the end index in inclusive. Compare it with\n",
    "# the result of previous examples.\n",
    "\n",
    "# Get a letter from the sequence (DNASeq object) at index 1.\n",
    "letter = seq[1]\n",
    "\n",
    "# Print the type of the returned object and the object itself\n",
    "# as if it was a string variable:\n",
    "print(type(letter), letter, sep='\\n', end='\\n\\n')\n",
    "\n",
    "# Get a slice of the DNASeq object (should be another DNASeq object):\n",
    "s = seq[4:9]\n",
    "\n",
    "# Print the type of the returned object and its content\n",
    "# as if it was a string variable:\n",
    "print(type(s), s, sep='\\n', end='\\n\\n')\n",
    "\n",
    "# In order to get a DNASeq object containing one letter sequence,\n",
    "# you may get a slice of length 1, using indexing compliant\n",
    "# with the GenBank notation:\n",
    "\n",
    "# Get a slice of the DNASeq object of length 1:\n",
    "s = seq[1:1]\n",
    "\n",
    "# Check out its type and print it as if it was a string variable:\n",
    "print(type(s), s, sep='\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gmk_loc(4_9) Guanylate kinase\n",
      "ATATTT\n",
      "\n",
      ">gmk_loc(4_9)_revcmpl Guanylate kinase\n",
      "AAATAT\n",
      "\n",
      ">gmk_loc(4_9)_revcmpl Guanylate kinase\n",
      "AAATAT\n"
     ]
    }
   ],
   "source": [
    "# Since a slice of DNASeq object is still a DNASeq object,\n",
    "# and a reverse complement of a DNASeq object is still\n",
    "# a DNASeq object, you may work with them like\n",
    "# with any DNASeq object.\n",
    "# Here you will print a fragment of the original sequence\n",
    "# as well as its reverse complement as a FASTA-formated string,\n",
    "# thanks to having the special __str__() method implemented:\n",
    "\n",
    "print(seq[4:9], end='\\n\\n')\n",
    "\n",
    "print(seq[4:9].revcmpl(), end='\\n\\n')\n",
    "\n",
    "# The reverse complementary sequence you can also obtained\n",
    "# by slicing a DNASeq object directly with the start index\n",
    "# value higher than the one for the end index, thanks to\n",
    "# having the special method __getitem__() implemented.\n",
    "\n",
    "print(seq[9:4])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 9<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "&gt;arcC_loc(1_10) Carbamate kinase<br>\n",
    "TTATTAATCC<br><br>\n",
    "&gt;glpF_loc(1_10) Glycerol kinase<br>\n",
    "GGTGCTGATT<br><br>\n",
    "&gt;arcC_loc(1_10)_glpF_loc(1_10) Carbamate kinase_Glycerol kinase<br>\n",
    "TTATTAATCCGGTGCTGATT\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">arcC_loc(1_10) Carbamate kinase\n",
      "TTATTAATCC\n",
      "\n",
      ">glpF_loc(1_10) Glycerol kinase\n",
      "GGTGCTGATT\n",
      "\n",
      ">arcC_loc(1_10)_glpF_loc(1_10) Carbamate kinase_Glycerol kinase\n",
      "TTATTAATCCGGTGCTGATT\n"
     ]
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# Select 10 nucleotide fragments of two different\n",
    "# sequences and add them together:\n",
    "seq_1 = seqs['arcC'][1:10]\n",
    "\n",
    "seq_2 = seqs['glpF'][1:10]\n",
    "\n",
    "seq_3 = seq_1 + seq_2\n",
    "\n",
    "# Print the initial sequences and the result of their addition\n",
    "# as if they were string values:\n",
    "print(seq_1, seq_2, seq_3, sep='\\n\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "<p style=\"background:#ffeedd;font-family:Sans;font-size:12pt;font-weight:bold;padding:10px;margin:0px\">\n",
    "Section 10<br><br>\n",
    "<span style=\"font-size:10pt\">The output should be:</span>\n",
    "</p>\n",
    "<p style=\"background:#ddffee;font-family:Monospace;font-size:10pt;padding:10px;margin:0px\">\n",
    "pta<br><br>\n",
    "new object<br><br>\n",
    "===<br><br>\n",
    "pta<br><br>\n",
    "new ref. the same obj.<br><br>\n",
    "new ref. the same obj.\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pta\n",
      "\n",
      "new object\n",
      "\n",
      "===\n",
      "\n",
      "pta\n",
      "\n",
      "new ref. the same obj.\n",
      "\n",
      "new ref. the same obj.\n"
     ]
    }
   ],
   "source": [
    "# Reload the sequences to have a collection of objects\n",
    "# that are instances of the up-to-date DNASeq class:\n",
    "\n",
    "seqs = DNASeq.from_file('input/Staphylococcus_MLST_genes.fasta')\n",
    "\n",
    "# Select one of the sequences by its sequence id (seqid):\n",
    "seq = seqs['pta']\n",
    "\n",
    "# Our method copy() allows to create\n",
    "# an independent copy of an existing object:\n",
    "\n",
    "seq_copy = seq.copy()\n",
    "\n",
    "# An object copy is an independent object,\n",
    "# stored independently in the RAM memory.\n",
    "# You can check it by changing seqid of the copy\n",
    "# and comparing it with seqid of the initial object:\n",
    "seq_copy.seqid = 'new object'\n",
    "\n",
    "print(seq.seqid, seq_copy.seqid, sep='\\n\\n', end='\\n\\n===\\n\\n')\n",
    "\n",
    "# This is in contrast to the operation of assignment,\n",
    "# which simply creates a new reference to\n",
    "# the same, already existing in memory object.\n",
    "# No you will only have two references leading\n",
    "# to the same address in the RAM memory.\n",
    "# You can test it using the code below and comparing\n",
    "# its results to the previous results obtained for\n",
    "# the implemented copy() method.\n",
    "\n",
    "print(seq.seqid, end='\\n\\n')\n",
    "\n",
    "# That line does not create an independent object,\n",
    "# it only creates yet another reference to the same\n",
    "# address in the RAM memory:\n",
    "new_ref = seq\n",
    "\n",
    "# Check it out for yourself:\n",
    "new_ref.seqid = 'new ref. the same obj.'\n",
    "\n",
    "print(seq.seqid, new_ref.seqid, sep='\\n\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddeeff;border-radius:15px\">The End :)</p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
